### PR TITLE
Week 7 makemehapi

### LINKED ISSUE #211: link to an instructions file, which references the original makemehapi repo
I know that the week 7 mentors were considering making changes to some of the exercises, but it doesn't look like that has happened?

We can still raise issues on the original, and see whether Matt Hernandez is accepting PRs. But in the meantime, let's just link directly to the original repo? It would make it much easier for students to submit issues, and I'm not sure what the forked version adds?

https://github.com/foundersandcoders/master-reference/blob/master/coursebook/week-7/README.md
https://github.com/msachi/makemehapi

### PR DESCRIPTION
**Pull request by msachi**
Fixes #211

### CODE REVIEW

File: coursebook/week-7/makemehapi-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
+## Make Me Hapi
+
+**makemehapi** is a workshop that will run you through a series of challenges ranging from a basic "hello world" server to more advanced exercises dealing with rendering views and helpers functions.
+
+The official workshop lives in this [GitHub repo](https://github.com/msachi/makemehapi), however, it is a little confusing and the wording of exercises isn't always great. To help you out, we've written the guidelines below -- if you follow these you should be fine!
```
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: link should be to the original repo

Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: We could make this slightly more positive. Nothing you're saying is incorrect :wink: 
But think about this from the point of view of someone who has never encountered the tutorial before. There's a reason we're still using it. It is good. We're just providing extra help

File: coursebook/week-7/makemehapi-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+To get set up:  
+1) Install makemehapi globally by running `npm install -g makemehapi`  
+2) Create a folder on your computer, e.g. `makemehapi-workshop` and `cd` into it  
+3) Run `npm init` to initialise a `package.json`, and install `hapi`. You will use this one and only `package.json` to install additional plugins as specified in the exercises.
+
+For each exercise:  
+1) Create a new subfolder, e.g. `exercise-1` and `cd` into it  
+2) Follow the instructions below for that exercise  
+3) To test your code:  
+~ run `makemehapi` in the Terminal
```
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: The `makemehapi` command takes them back to the list of exercises.

To test your code, type the following command:
`makemehapi run <<name of your file>>`

File: coursebook/week-7/makemehapi-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+1) Install makemehapi globally by running `npm install -g makemehapi`  
+2) Create a folder on your computer, e.g. `makemehapi-workshop` and `cd` into it  
+3) Run `npm init` to initialise a `package.json`, and install `hapi`. You will use this one and only `package.json` to install additional plugins as specified in the exercises.
+
+For each exercise:  
+1) Create a new subfolder, e.g. `exercise-1` and `cd` into it  
+2) Follow the instructions below for that exercise  
+3) To test your code:  
+~ run `makemehapi` in the Terminal  
+~ navigate to the relevant exercise using the command-line tool
```
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: not necessary if they specify the full path name in the next step

File: coursebook/week-7/makemehapi-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+The official workshop lives in this [GitHub repo](https://github.com/msachi/makemehapi), however, it is a little confusing and the wording of exercises isn't always great. To help you out, we've written the guidelines below -- if you follow these you should be fine!
+
+To get set up:  
+1) Install makemehapi globally by running `npm install -g makemehapi`  
+2) Create a folder on your computer, e.g. `makemehapi-workshop` and `cd` into it  
+3) Run `npm init` to initialise a `package.json`, and install `hapi`. You will use this one and only `package.json` to install additional plugins as specified in the exercises.
+
+For each exercise:  
+1) Create a new subfolder, e.g. `exercise-1` and `cd` into it
```
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: Having subfolders is probably overkill? How about just `ex1.js`, `ex2.js`, etc?

File: coursebook/week-7/makemehapi-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+For each exercise:  
+1) Create a new subfolder, e.g. `exercise-1` and `cd` into it  
+2) Follow the instructions below for that exercise  
+3) To test your code:  
+~ run `makemehapi` in the Terminal  
+~ navigate to the relevant exercise using the command-line tool  
+~ run `makemehapi verify path-to-file` (e.g. `makemehapi verify server.js`, provided you are already in `exercise-1` subfolder)  
+
+### HELLO HAPI
```
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: Is all of this copied from the tutorial itself, or do you provide extra hints that aren't in `makemehapi` already?

If it's just a copy of the tutorial...I'm not sure whether we need it. Otherwise we have to continually update this file if/when the tutorial changes.

If the instructions are identical, we might as well just tell them to start following the `makemehapi` tutorial itself, in the command line? Otherwise, students might wonder whether they're supposed to read through this file, or look at their terminal

File: coursebook/week-7/makemehapi-guidelines.md (hunk 1 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: oops!

Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: I thought I was relatively positive ;) I'll rephrase

File: coursebook/week-7/makemehapi-guidelines.md (hunk 4 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: It's not always just one file though. Sometimes there's an extra html to serve, and sometimes even more files (e.g. the helpers exercise - a separate file with the helpers function). Would get very confusing to have all that stuff from separate exercises in one folder.

File: coursebook/week-7/makemehapi-guidelines.md (hunk 3 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: Even if they specify the full path, the program needs to know what solution it's testing against. And I think you can only tell it by navigating to that exercise using the old-school gui.

File: coursebook/week-7/makemehapi-guidelines.md (hunk 2 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: See below.

File: coursebook/week-7/makemehapi-guidelines.md (hunk 5 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: Definitely not identical! That's what I spent most of the afternoon doing - rephrasing what was unclear, and adding an extra sentence here and there to make it clearer.

**Comment by jsms90** [comment]
@msachi let me know if you're no longer available to make these changes before tomorrow?

**Comment by msachi** [comment]
@jsms90 I can make them now just review the comments above

File: coursebook/week-7/makemehapi-guidelines.md (hunk 3 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: oh! I thought you mean the relevant exercise folder that the students have created, using the command line to `cd` into the correct folder

File: coursebook/week-7/makemehapi-guidelines.md (hunk 5 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: ahhhhh I see

hmmm...so should they be following your makemehapi-instructions.md _instead_ of the tutorial itself? To be honest, I'm not sure that that will happen. If they have to use the tutorial itself to run files and verify files, they're probably going to read the instructions that are given there too.

But it might be useful to provide some points of clarification, that assume they have just read the tutorial itself.

If it looks very similar, I'm afraid that they might end up not reading yours? :thinking:

File: coursebook/week-7/makemehapi-guidelines.md (hunk 3 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: Ah okay. Let me rephrase.

File: coursebook/week-7/makemehapi-guidelines.md (hunk 5 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: They don't have to run files using `makemehapi`, that's why I omitted that instruction. I think it's pretty clear from the guidelines what they should be doing - i.e. using `makemehapi` just for testing. I can add another sentence to make it even more obvious, and stress it in the workshop.

It's small things that make a difference. It would not make sense to say to them, you know that word at the end of the second sentence of exercise 2. That should be 'path' instead of 'route'. Just replacing the whole thing is the best approach IMO. Obviously ideally the new instructions would be part of `makemehapi` itself. But seeing as it's 21.30 and Nazareth need this tomorrow, I think we should leave it as it is.

File: coursebook/week-7/makemehapi-guidelines.md (hunk 4 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: Yeah it's true. But there's very little else that repeats. As far as I can see, there's only one other duplication, and that's two file inside the `templates` folder. I reckon they can handle making `index1.html` and `index2.html`, and it's probably simpler than having a folder for each different exercise

File: coursebook/week-7/makemehapi-guidelines.md (hunk 5 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: Yeah I do understand what you mean. Obviously it would be better to just clarify the content of the workshop.

I only see this as an issue, if they don't end up reading the instructions in this file, once they start the tutorial. Or if it causes confusion that there are two sets of instructions. But your re-wording may be enough:
> Note: The instructions below are a modified version of those you will see in your Terminal

:+1:

File: coursebook/week-7/makemehapi-guidelines.md (hunk 4 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: It may be more confusing to have to re-write
> path: Path.join(__dirname, 'exercise-5', 'templates')

to
> path: Path.join(__dirname, 'exercise-7', 'templates')

especially since they're likely to be copy-pasting whatever they can from a previous exercise

File: coursebook/week-7/makemehapi-guidelines.md (hunk 2 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: ohhhh

Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: If you're going to add the commands in here, do add `makemehapi run path-to-file` though.

It's good to remind them that they can use this to see their output, rather than relying on `verify` all the time, which doesn't usually provide any useful feedback for incorrect results

File: coursebook/week-7/makemehapi-guidelines.md (hunk 4 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: Personally I would find it more confusing to have to think about whether I'm serving this index.html or that index.html. And I think it's good practice to have one folder per app / back-end, mixing them up seems heresy to me. As for path.joining, there's no need to do that, everything will be relative to the folder you're in, so just `path.join(__dirname, 'templates')`

File: coursebook/week-7/makemehapi-guidelines.md (hunk 2 above)
Comment (msachi) [inline coursebook/week-7/makemehapi-guidelines.md]: My thinking was just that `makemehapi run server.js` doesn't do anything more than `node sever.js`, so no point introducing an additional redundant command?

**Comment by esraajb** [comment]
I'm going to merge it as we need to use it now

File: coursebook/week-7/makemehapi-guidelines.md (hunk 2 above)
Comment (jsms90) [inline coursebook/week-7/makemehapi-guidelines.md]: Fair enough. The only nice thing about the `makemhapi run` command is that it allows you to stay in your terminal. Less switching around. But it's actually probably much better practice to just use `node`!
